### ISSUE TITLE
Segmentation fault with bun run --hot

### ISSUE THREAD

**Issue by RedEagle-dh** [oven-sh/bun#10274]
### What version of Bun is running?

1.1.3

### What platform is your computer?

Microsoft Windows NT 10.0.22631.0 x64

### What steps can reproduce the bug?

# Create a typescript discord.js bot with bun
## Install the following dependencies
```json
{
  "dependencies": {
    "@discordjs/rest": "^2.2.0",
    "discord.js": "^14.14.1",
    "dotenv": "^16.4.5",
    "pino": "^8.20.0",
    "pino-pretty": "^11.0.0"
  },
  "devDependencies": {
    "@types/node": "^20.12.7",
    "nodemon": "^3.1.0",
    "prettier": "^3.2.5",
    "typescript": "^5.4.5"
  }
}
```

### tsconfig.json
```ts
{
	"compilerOptions": {
		"target": "ES6",
		"module": "CommonJS",
		"rootDir": ".",
		"baseUrl": ".",
		"outDir": "./dist",
		"esModuleInterop": true,
		"skipLibCheck": true,
		"forceConsistentCasingInFileNames": true,
		"strict": false,
		"experimentalDecorators": true,
		"emitDecoratorMetadata": true,
		"paths": {
			"@/*": [
				"./*"
			],
			"@lib/*": [
				"./src/lib/*"
			]
		}
	},
	"include": [
		"**/*.ts"
	],
	"exclude": [
		"node_modules"
	]
}
```

### src/index.ts
```ts
import getDiscordClient, { login, registerEvents } from './client';
import { deployCommands } from './setupCommands/deployCommands';
import { registerCommands } from './setupCommands/registerCommands';
import { Logger } from '@lib/classes/Logger';

const log = new Logger('GEE Bot');

const client = getDiscordClient();

export const returnClient = () => {
	return client;
};

const startBot = async () => {
	const commands = await deployCommands(log);
	registerCommands(client, log, commands);
	registerEvents(log);

	login(log);
};

startBot().catch((error) => {
	log.error(error.stack || error);
});
```

### src/client.ts

```ts
import { Client, GatewayIntentBits, Partials, Collection } from 'discord.js';

import * as events from './events';
import { Logger } from './lib/classes/Logger';

declare module 'discord.js' {
	interface Client {
		commands: Collection<string, Interaction>;
	}
}

function getDiscordClient() {
	client.setMaxListeners(19);
	return client;
}

export const client = new Client({
	intents: [
		GatewayIntentBits.DirectMessages,
		GatewayIntentBits.GuildMembers,
		GatewayIntentBits.GuildMessages,
		GatewayIntentBits.GuildScheduledEvents,
		GatewayIntentBits.GuildPresences,
		GatewayIntentBits.MessageContent,
		GatewayIntentBits.DirectMessageTyping,
		GatewayIntentBits.GuildVoiceStates,
		GatewayIntentBits.Guilds,
		GatewayIntentBits.GuildMessageReactions,
	],
	partials: [
		Partials.GuildScheduledEvent,
		Partials.GuildMember,
		Partials.Message,
		Partials.User,
		Partials.Reaction,
	],
});

export function registerEvents(__LOG: Logger) {
	__LOG.info('Started loading events...');
	const eventModules = Object.values(events);
	for (const eventModule of eventModules) {
		const eventKeys = Object.keys(eventModule);

		for (const eventName of eventKeys) {
			const event = eventModule[eventName];
			const eventExecute = event.execute;
			__LOG.debug(`Loading event [${event.name}]`);
			if (event.once) {
				client.once(event.name, (...args) =>
					eventExecute(...[...args, ...[__LOG, client]]),
				);
			} else {
				client.on(event.name, (...args) =>
					eventExecute(...[...args, ...[__LOG, client]]),
				);
			}
		}
	}
	__LOG.info(`${eventModules.length} events loaded!`);
}

export function login(__LOG: Logger) {
	__LOG.info('Logging in...');
	client.login(process.env.DISCORD_TOKEN || '').catch((err) => {
		__LOG.error('Login failed: ' + err);
	});
}

export default getDiscordClient;
```

### src/setupCommands/deployCommands.ts

```ts
import * as commandModules from '@/src/commands';

import { SlashCommandBuilder, REST, Routes } from 'discord.js';

import type { Command } from '@lib/types/Command';
import { Logger } from '../lib/classes/Logger';

const rest = new REST().setToken(`${process.env.DISCORD_TOKEN}`);

export async function deployCommands(log: Logger): Promise<Command[]> {
	const commands: SlashCommandBuilder[] = [];
	const returnObject: Command[] = [];

	for (const module of Object.values<Command>(commandModules)) {
		returnObject.push(module);
		log.debug(`Loaded command: ${JSON.stringify(module.data)}`);
		commands.push(module.data);
	}

	log.warn(`Refreshing ${commands.length} application (/) commands...`);
	await rest
		.put(Routes.applicationCommands(`${process.env.APPLICATION_ID}`), {
			body: commands,
		})
		.then(() => {
			log.info(
				`${commands.length} application (/) commands loaded successfully.`,
			);
		})
		.catch((error) => {
			log.error(
				'Failed to parse token in ' +
					__filename.split('\\').reverse()[0],
			);
			console.log(error);
			process.exit();
		});
	return returnObject;
}

export default deployCommands;
```

### src/setupCommands/registerCommands.ts

```ts
import { Logger } from '../lib/classes/Logger';
import { Command } from '../lib/types/Command';
import { Client, Collection } from 'discord.js';

export function registerCommands(
	client: Client,
	__LOG: Logger,
	commands: Command[],
) {
	client.commands = new Collection();
	for (const command of commands) {
		// @ts-ignore
		client.commands.set(command.data.name, command);
	}
}
```

### src/lib/classes/Logger.ts

```ts
import createLogger, { multistream, transport } from 'pino';
import type { Logger as PinoLogger } from 'pino';
import type { PrettyOptions } from 'pino-pretty';

const pinoPrettyOptions: PrettyOptions = {
	colorize: true,
	levelFirst: true,
	translateTime: true,
};

export class Logger {
	public readonly pino: PinoLogger;

	public constructor(private readonly name: string) {
		this.pino = createLogger(
			{
				name: this.name,
				level: process.env.NODE_ENV === 'production' ? 'info' : 'trace',
			},
			multistream([
				{
					stream: transport({
						target: 'pino-pretty',
						options: pinoPrettyOptions,
					}),
					level:
						process.env.NODE_ENV === 'production'
							? 'info'
							: 'trace',
				},
			]),
		);
	}

	public trace(message: string): void {
		this.pino.trace(message);
	}

	public debug(message: string): void {
		this.pino.debug(message);
	}

	public info(message: string): void {
		this.pino.info(message);
	}

	public warn(message: string): void {
		this.pino.warn(message);
	}

	public error(message: string): void {
		this.pino.error(message);
	}

	public fatal(message: string): void {
		this.pino.fatal(message);
	}
}
```

### src/lib/types/Command.ts

```ts
import { CommandInteraction, SlashCommandBuilder } from 'discord.js';
export type Command = {
	data: SlashCommandBuilder;
	execute: (event: CommandInteraction) => void;
};
```

### src/commands/util/say.ts

```ts
import {
	SlashCommandBuilder,
	CommandInteraction,
	TextChannel,
} from 'discord.js';

export const data = new SlashCommandBuilder()
	.setName('say')
	.setDescription('Sends a message in this channel.')
	.addStringOption((option) =>
		option
			.setName('message')
			.setDescription('The message you want to send.')
			.setRequired(true),
	)
	.setDMPermission(false);

export async function execute(event: CommandInteraction) {
	const message = event.options.get('message')?.value as string;
	if (!message) {
		await event.reply({
			content: 'Error.',
			ephemeral: true,
		});
	}

	const channel = event.channel as TextChannel;
	await event.reply({
		content: 'OK',
		ephemeral: true,
	});

	await channel.send(message);
}
```

### src/commands/index.ts

```ts
export * as say from './util/say';
```

### src/events/interaction/interactionCreate.ts

```ts
import { InteractionType, Events, Interaction } from 'discord.js';

import { handleMessageComponentInteraction } from './MessageComponents';
import { handleModalSubmitInteraction } from './ModalSubmit';
import { Logger } from '@/src/lib/classes/Logger';

export const interactionCreate = {
	name: Events.InteractionCreate,
	once: false,
	async execute(event: Interaction, __LOG: Logger) {
		__LOG.debug('Interaction event fired');
		switch (event.type) {
			case InteractionType.ApplicationCommand: {
				const { commandName } = event;
				const command = event.client.commands.get(commandName);
				if (!command) return;

				try {
					__LOG.info(`${event.user.tag} used /${event.commandName}`);
					// @ts-ignore
					await command.execute(event, __LOG);
				} catch (error: any) {
					__LOG.error(error.stack || error);
				}
				break;
			}
		}
	},
};
```

### src/events/ready/ready.ts
```ts
import { Logger } from '@/src/lib/classes/Logger';
import {
	Client,
	ActivityType,
	ActivitiesOptions,
	Events,
	Guild,
} from 'discord.js';

const activity: ActivitiesOptions[] = [
	{ name: "TikToks", type: ActivityType.Watching },
];

export const clientReady = {
	name: Events.ClientReady,
	once: true,
	async execute(client: Client, __LOG: Logger) {
		__LOG.info(
			`Logged in as ${client.user?.tag} on ${
				client.guilds.cache.size
			} Servers: ${client.guilds.cache
				.map((g: Guild) => g.name)
				.join(', ')}`,
		);
		client.user?.setPresence({
			activities: activity,
		});
		__LOG.info(
			`Activity of ${client.user?.tag} is ${client.user?.presence.activities}`,
		);
	},
};
```

### src/events/index.ts

```ts
export * from './ready';
export * from './interaction';
```

### src/events/ready/index.ts

```ts
export * as ready from "./ready";
```

### src/events/interaction/index.ts

```ts
export * as interactionCreate from "./interactionCreate";
```

### .env
You can create a bot in the [discord developer portal](https://discord.com/developers/applications)
```env
DISCORD_TOKEN=YOUR BOT TOKEN
NODE_ENV=development
APPLICATION_ID=YOUR CLIENT ID
```

## Run the following command

```bash
bun run --hot src/index.ts
```

### What is the expected behavior?

It should restart the bot, reload, redeploy and reregister all commands and events.

### What do you see instead?

After reloading all events and commands, when the bot is logging in, a segmentation fault is happening at address `0xffffffffffffffff`.

You can see, it starts successfully and restarts after `Message created: Hi!`.

```shell
$ bun run --hot src/index.ts
DEBUG [13:37:09.442] (Bot/6576): Loaded command: {"options":[{"type":3,"name":"message","description":"Die Nachricht die zu senden möchtest.","required":true}],"name":"say","description":"Sendet eine Nachricht in diesen Channel.","dm_permission":false}
WARN [13:37:09.442] (Bot/6576): Refreshing 1 application (/) commands...
INFO [13:37:09.709] (Bot/6576): 1 application (/) commands loaded successfully.
INFO [13:37:09.709] (Bot/6576): Started loading events...
DEBUG [13:37:09.709] (Bot/6576): Loading event [channelCreate]
DEBUG [13:37:09.709] (Bot/6576): Loading event [channelDelete]
DEBUG [13:37:09.709] (Bot/6576): Loading event [channelUpdate]
DEBUG [13:37:09.709] (Bot/6576): Loading event [guildMemberAdd]
DEBUG [13:37:09.709] (Bot/6576): Loading event [guildMemberRemove]
DEBUG [13:37:09.709] (Bot/6576): Loading event [guildMemberUpdate]
DEBUG [13:37:09.709] (Bot/6576): Loading event [guildScheduledEventCreate]
DEBUG [13:37:09.709] (Bot/6576): Loading event [guildScheduledEventDelete]
DEBUG [13:37:09.709] (Bot/6576): Loading event [guildScheduledEventUpdate]
DEBUG [13:37:09.709] (Bot/6576): Loading event [guildScheduledEventUserAdd]
DEBUG [13:37:09.709] (Bot/6576): Loading event [guildScheduledEventUserRemove]
DEBUG [13:37:09.709] (Bot/6576): Loading event [interactionCreate]
DEBUG [13:37:09.709] (Bot/6576): Loading event [messageCreate]
DEBUG [13:37:09.709] (Bot/6576): Loading event [messageReactionAdd]
DEBUG [13:37:09.709] (Bot/6576): Loading event [ready]
DEBUG [13:37:09.709] (Bot/6576): Loading event [roleCreate]
DEBUG [13:37:09.709] (Bot/6576): Loading event [roleDelete]
DEBUG [13:37:09.709] (Bot/6576): Loading event [roleUpdate]
DEBUG [13:37:09.709] (Bot/6576): Loading event [voiceStateUpdate]
INFO [13:37:09.709] (Bot/6576): 19 events loaded!
INFO [13:37:09.709] (Bot/6576): Logging in...
INFO [13:37:10.404] (Bot/6576): Logged in as Bot#2860 on 1 Servers: Bot Developing
INFO [13:37:10.404] (Bot/6576): Activity of Bot#2860 is TikToks
DEBUG [13:37:14.943] (Bot/6576): Interaction event fired
INFO [13:37:14.943] (Bot/6576): redeagle. used /say
DEBUG [13:37:15.186] (Bot/6576): Message created: OK
DEBUG [13:37:15.404] (Bot/6576): Message created: Hi!
DEBUG [13:37:25.572] (Bot/6576): Loaded command: {"options":[{"type":3,"name":"message","description":"Die Nachricht die zu senden möchtest.","required":true}],"name":"say","description":"Sendet eine Nachricht in diesen Channel.","dm_permission":false}
WARN [13:37:25.573] (Bot/6576): Refreshing 1 application (/) commands...
DEBUG [13:37:25.833] (Bot/6576): Loaded command: {"options":[{"type":3,"name":"message","description":"Die Nachricht die zu senden möchtest.","required":true}],"name":"say","description":"Sendet eine Nachricht in diesen Channel.","dm_permission":false}
WARN [13:37:25.833] (Bot/6576): Refreshing 1 application (/) commands...
INFO [13:37:26.051] (Bot/6576): 1 application (/) commands loaded successfully.
INFO [13:37:26.051] (Bot/6576): Started loading events...
DEBUG [13:37:26.052] (Bot/6576): Loading event [channelCreate]
DEBUG [13:37:26.052] (Bot/6576): Loading event [channelDelete]
DEBUG [13:37:26.052] (Bot/6576): Loading event [channelUpdate]
DEBUG [13:37:26.052] (Bot/6576): Loading event [guildMemberAdd]
DEBUG [13:37:26.052] (Bot/6576): Loading event [guildMemberRemove]
DEBUG [13:37:26.052] (Bot/6576): Loading event [guildMemberUpdate]
DEBUG [13:37:26.052] (Bot/6576): Loading event [guildScheduledEventCreate]
DEBUG [13:37:26.052] (Bot/6576): Loading event [guildScheduledEventDelete]
DEBUG [13:37:26.052] (Bot/6576): Loading event [guildScheduledEventUpdate]
DEBUG [13:37:26.052] (Bot/6576): Loading event [guildScheduledEventUserAdd]
DEBUG [13:37:26.052] (Bot/6576): Loading event [guildScheduledEventUserRemove]
DEBUG [13:37:26.052] (Bot/6576): Loading event [interactionCreate]
DEBUG [13:37:26.052] (Bot/6576): Loading event [messageCreate]
DEBUG [13:37:26.052] (Bot/6576): Loading event [messageReactionAdd]
DEBUG [13:37:26.052] (Bot/6576): Loading event [ready]
DEBUG [13:37:26.052] (Bot/6576): Loading event [roleCreate]
DEBUG [13:37:26.052] (Bot/6576): Loading event [roleDelete]
DEBUG [13:37:26.052] (Bot/6576): Loading event [roleUpdate]
DEBUG [13:37:26.052] (Bot/6576): Loading event [voiceStateUpdate]
INFO [13:37:26.052] (Bot/6576): 19 events loaded!
INFO [13:37:26.052] (Bot/6576): Logging in...
Segmentation fault at address 0xffffffffffffffff
???:?:?: 0x7ff6a99258c6 in ??? (bun.exe)
???:?:?: 0x7ff6a80be3ed in ??? (bun.exe)
???:?:?: 0x7ff6a80b15db in ??? (bun.exe)
???:?:?: 0x7ff6a80b102f in ??? (bun.exe)
???:?:?: 0x7ff6a7465b35 in ??? (bun.exe)
???:?:?: 0x7ff6a98be5da in ??? (bun.exe)
???:?:?: 0x7ff6a739d7aa in ??? (bun.exe)
???:?:?: 0x7ff6a73a097a in ??? (bun.exe)
???:?:?: 0x7ff6a73b46eb in ??? (bun.exe)
???:?:?: 0x7ff6a70e1603 in ??? (bun.exe)
???:?:?: 0x7ff6a9c83eb3 in ??? (bun.exe)
???:?:?: 0x7ffcb703257c in ??? (KERNEL32.DLL)
???:?:?: 0x7ffcb87eaa47 in ??? (ntdll.dll)
```

### Additional information

The bug also occurred when I started the bot for the first time without restarting it or making use of the hot reload.
Without --hot everything works perfectly.

**Comment by depthbomb** [comment]
Seems to happen while the library is connecting to Discord, as seen in this basic reproduction:

![rh3lw5](https://github.com/oven-sh/bun/assets/6052766/79434232-d981-4932-bc80-1b18762b0b98)

**package.json**

```json
{
	"name": "djs-repro",
	"type": "module",
	"scripts": {
		"start": "bun run src/index.ts"
	},
	"dependencies": {
		"bufferutil": "^4.0.8",
		"discord.js": "^14.14.1",
		"utf-8-validate": "^6.0.3",
		"zlib-sync": "^0.1.9"
	},
	"devDependencies": {
		"@types/bun": "latest"
	},
	"peerDependencies": {
		"typescript": "^5.0.0"
	}
}
```

**src/index.ts**

```ts
import { Client, GatewayIntentBits } from 'discord.js';

const client = new Client({ intents: [GatewayIntentBits.Guilds] });

client.on('debug', msg => console.log(msg));

await client.login(process.env.BOT_TOKEN!);
```

**Comment by Jarred-Sumner** [comment]
This Discord.js error was fixed in Bun v1.1.10 or so when we addressed issues with websocket connection callback lifetimes
